keps/sig-api-machinery/3945-crd-conversion-with-cel/README.md

```diff
@@ -0,0 +1,912 @@
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. 
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+The [Common Expression Language (CEL)](https://github.com/google/cel-go) would be used for version conversion which
+currently can only be done through conversion webhooks.
+
+Current the conversion is done through the conversion strategy in CRD which supports 'none' and 'webhook'. 
+This KEP proposes to add `x-kubernetes-conversion` to CRD structural schemas to allow CEL conversion.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          type: object
+          properties:
+            demoField:
+              type: integer
+              x-kubernetes-conversion:
+                - from: v1 # from is the source version.
+                  to: v2 # to is the target version.
+                  configuration: # configuration contains field-specific conversion expressions.
+                    newFieldName: {$: "v1.spec.oldFieldName"}
+            ...
+```
+
+A new extension `x-kubernetes-conversion` is added to instruct API server to handle version conversion between custom resources.
+
+conversionRules is to specify rules to convert between different API versions:
+- `from` is the source version. 
+- `to` is the target version.
+- `configuration` contains field-specific conversion expressions.
+
+Notes: 
+- All fields which do not have `x-kubernetes-conversion` specified will be simply copied over.
+- Since `x-kubernetes-conversion` is embedded together with the field, if the field is unset, it will be ignored.
+
+[UNSOLVED] metadata access: conversion webhook have access to labels and annotation. 
+validation rules have access to name, generatedName, apiVersion, kind.
+`x-kubernetes-conversion` shouldn't have access to version info, what about others?
+
+TODO: we would love to have a unified behavior across CEL mutation use case hence the detail will be synced in a separate doc 
+and will be updated later.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+Use cases plan to support:
```

Comment:
Sorry, what I had meant to indicate with this list is that we need to assess the design proposal and see which of these might possibly work. It's fine to write them down as well, but the primary concern I had was that I wasn't convinced that the design had been checked against a sufficient set of use cases. I think that needs to happen.

Reply:
For example, showing how each of these cases would work with the proposed design would go a long way to ensuring we have something that is viable.